src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ExternalProcessRunner.java

```diff
@@ -0,0 +1,103 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.service;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+@Slf4j
+@Component
+public class ExternalProcessRunner {
+    /**
+     * Executes the given command with given arguments.
+     * @param command the command to execute
+     * @param args arguments to be appended to the command. Make sure to pass your arguments in the correct order
+     * (e.g. if your options have values enter them as separate consecutive args).
+     * @return the output of the executed process as a List of Strings
+     * @throws ProcessNotExecutableException in the case of IOException or if the process is interrupted
+     * @throws ProcessFailedException if the process' exit code is not 0
+     */
+    public List<String> execute(String command, String... args) throws ProcessNotExecutableException,
+            ProcessFailedException {
+        if (StringUtils.isEmpty(command)) {
+            throw new IllegalArgumentException("command cannot be null");
+        }
+
+        List<String> commandWithArgs = new ArrayList<>();
+        commandWithArgs.add(command);
+        if (args != null && args.length > 0) {
+            commandWithArgs.addAll(Arrays.asList(args));
+        }
+        Process process;
+        ProcessBuilder pb = new ProcessBuilder(commandWithArgs);
+        // redirect process errors into process's input stream
+        pb.redirectErrorStream(true);
+        try {
+            process = pb.start();
+        } catch (IOException e) {
+            throw new ProcessNotExecutableException(e);
+        }
+
+        // gather output into a list of string for returning
+        List<String> processOutput = new ArrayList<>();
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+            br.lines().forEach(processOutput::add);
+        } catch (IOException e) {
+            process.destroy();
+            throw new ProcessNotExecutableException(e);
+        }
+
+        int exitCode;
+
+        try {
+            exitCode = process.waitFor();
+        } catch (InterruptedException e) {
+            // we don't want to throw the InterruptedException from here but we want to retain the interrupted status
```

Comment:
Good that there's the `Thread.currentThread().interrupt();` which is important. 

I am a bit unsure about wrapping it in ProcessNotExecutableException though. Would just throwing InterruptedException from the methods be better? InterruptedException has a special meaning and wrapping it in ProcessNotExecutableException kind of hides this info from the callers. Declaring InterruptedException would probably be more "standard" / idiomatic way of handling this.  

For example https://stackoverflow.com/questions/3976344/handling-interruptedexception-in-java has lots of discussion which I kind of interpret advocating throwing InterruptedException unless there's special reason not to.

We would have to map InterruptedException -> InternalServerErrorException in the controller anyways, so the change would mostly affect the intermediate layers.

Reply:
Changed this so that the `InterruptedException` is propagated after `Thread.currentThread().interrupt();`. I also discussed this with @jhyoty and we concluded that `ProcessNotExecutableException` would be incorrect information to the end user because even though the executing thread itself is interrupted the process might be able to complete.

The controller layer now converts the exception to `InternalServerErrorException` and a proper code in the error response is attached (at the moment either `internal_key_cert_interrupted` in `SystemApiController` or `wsdl_validator_interrupted` when WSDL validation fails).